File: Leetcode Daily Challenge/027-September-2025/001-maximum-average-pass-ratio.java
// https://leetcode.com/problems/maximum-average-pass-ratio/




import java.util.*;

class Solution {
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<ClassInfo> pq = new PriorityQueue<>((a, b) -> 
            Double.compare(b.gain(), a.gain())
        );

        for (int[] cls : classes) {
            pq.offer(new ClassInfo(cls[0], cls[1]));
        }

        while (extraStudents -- > 0) {
            ClassInfo top = pq.poll();
            top.pass ++;
            top.total ++;
            pq.offer(top);
        }

        double totalRatio = 0.0;
        while (!pq.isEmpty()) {
            ClassInfo top = pq.poll();
            totalRatio += ((double)top.pass / top.total);
        }
        return (totalRatio / classes.length);
    }

    private static class ClassInfo {
        int pass, total;
        ClassInfo(int p, int t) {
            this.pass  = p;
            this.total = t;
        }

        double gain() {
            double before = (double) pass / total;
            double after  = (double) (pass + 1) / (total + 1);
            return (after - before);
        }
    }
}

File: Leetcode Daily Challenge/027-September-2025/001-maximum-average-pass-ratio.cpp
// https://leetcode.com/problems/maximum-average-pass-ratio/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    using T = pair<double, pair<int, int>>;

    // TC : O(N*log(N))
    // SC : O(N)
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        auto gain = [&](int pass, int total) {
            return ((double)(pass + 1) / (total + 1)) - ((double)pass / total);
        };

        priority_queue<T> pq;

        for (vector<int> &cls : classes) {
            int pass  = cls[0];
            int total = cls[1];
            pq.push({gain(pass, total), {pass, total}});
        }

        while (extraStudents --) {
            auto it = pq.top();
            pq.pop();

            double maxGain = it.first;
            int pass  = it.second.first;
            int total = it.second.second;

            pass ++, total ++;
            pq.push({gain(pass, total), {pass, total}});
        }

        double totalRatio = 0.0;
        while (!pq.empty()) {
            auto it = pq.top();
            pq.pop();
            
            int pass  = it.second.first;
            int total = it.second.second;
            totalRatio += ((double)pass / total);
        }

        return (totalRatio / classes.size());
    }
};